pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package appconfigreporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+
+	"github.com/pipe-cd/pipe/pkg/app/api/service/pipedservice"
+	"github.com/pipe-cd/pipe/pkg/cache"
+	"github.com/pipe-cd/pipe/pkg/cache/memorycache"
+	"github.com/pipe-cd/pipe/pkg/config"
+	"github.com/pipe-cd/pipe/pkg/git"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+const (
+	defaultLastFetchedCommitCacheSize = 500
+)
+
+type apiClient interface {
+	UpdateApplicationConfigurations(ctx context.Context, in *pipedservice.UpdateApplicationConfigurationsRequest, opts ...grpc.CallOption) (*pipedservice.UpdateApplicationConfigurationsResponse, error)
+	PutUnregisteredApplicationConfigurations(ctx context.Context, in *pipedservice.PutUnregisteredApplicationConfigurationsRequest, opts ...grpc.CallOption) (*pipedservice.PutUnregisteredApplicationConfigurationsResponse, error)
+}
+
+type gitClient interface {
+	Clone(ctx context.Context, repoID, remote, branch, destination string) (git.Repo, error)
+}
+
+type applicationLister interface {
+	List() []*model.Application
+}
+
+type Reporter struct {
+	apiClient         apiClient
+	gitClient         gitClient
+	applicationLister applicationLister
+	config            *config.PipedSpec
+	gitRepos          map[string]git.Repo
+	gracePeriod       time.Duration
+	// Cache for the last fetched commit for each repository.
+	lastFetchedCommitCache cache.Cache
```

Comment:
This is just for marking the hash value of the previously scanned commit, so to be simple we can use a normal map.

Reply:
Right